### PR TITLE
Merge memory regions before checking for reserved areas

### PR DESCRIPTION
**Pull request by bvirfollet**
As mentioned [here](https://sel4.discourse.group/t/elfloader-memory-mapping-boundaries), contiguous regions can lead to memory fragmentation. Merging those before checking for reserved area would prevent it.
Signed-off-by: Bertrand Virfollet <bvirfollet@silicom.fr>

### CODE REVIEW

File: tools/hardware/utils/memory.py [hunk 1]
```diff
@@ -26,6 +26,38 @@ def visitor(node: WrappedNode):
     return regions
 
 
+def merge_memory_regions(regions: Set[Region]) -> Set[Region]:
+    ''' Check all region and merge adjacent ones '''
+    all_regions = []
+    idx = 0
+    for region in regions:
```
Comment (oliver-wm) [inline tools/hardware/utils/memory.py:33]: prefer enumerate here

File: tools/hardware/utils/memory.py [hunk 2]
```diff
@@ -26,6 +26,38 @@ def visitor(node: WrappedNode):
...
+        all_regions.append(dict(idx=idx, region=region, right_adj=None, left_adj=None))
+        idx += 1
+
+    # Find all right contiguous regions
+    for dreg in all_regions:
+        for dnreg in all_regions[dreg['idx']+1:]:
```
Comment (oliver-wm) [inline tools/hardware/utils/memory.py:39]: This assume that all_regions is sorted?

File: tools/hardware/utils/memory.py (hunk 1 above)
Comment (bvirfollet) [reply tools/hardware/utils/memory.py:33]: What do you mean? Get everything done in this loop?

Comment (oliver-wm) [reply tools/hardware/utils/memory.py:33]:
```suggestion
    for idx, region in enumerate(regions):
```

Comment (oliver-wm) [reply tools/hardware/utils/memory.py:33]: Then you don't need to worry about incrementing or declaring the idx variable

Comment (bvirfollet) [reply tools/hardware/utils/memory.py:33]: Yeap, sorry I'm a bit old school on this :)

File: tools/hardware/utils/memory.py (hunk 2 above)
Comment (bvirfollet) [reply tools/hardware/utils/memory.py:39]: It seems important to keep separated the device memory regions and untyped memory regions. All memory regions are inspected, but contiguity is considered in respect to region owner. And as a result, a memory region spanning over 2 buses would not get merged.

**Review by kent-mcleod — APPROVED** [review]

**Review by nomadeel — APPROVED** [review]

**Comment by nomadeel** [comment]
Looks good, but I had a look at the commits and there's a merge commit. If possible, could you remove that commit and use rebase instead? So for your case, `git rebase origin/master` where origin is upstream.

**Comment by lsf37** [comment]
If the merge commit is from GitHub's `update branch` button, it will disappear if you do a "Rebase and merge".

**Comment by lsf37** [comment]
What is failing and will prevent the merge is `gitlint`, though (the title of the commit messages needs to be < 50 characters)

**Review by oliver-wm — APPROVED** [review]

**Comment by lsf37** [comment]
Merged via #562

**Pull request closed without merging**